File: Hash Table/0904.Fruit Into Baskets/Solution.java
class Solution {
    public int totalFruit(int[] fruits) {
        Map<Integer, Integer> cnt = new HashMap<>();
        int l = 0, r = 0, n = fruits.length;
        int ans = 0, cur = 0;
        while (r < n) {
            while (r < n && (cnt.size() < 2 || cnt.containsKey(fruits[r]))) {
                cnt.put(fruits[r], cnt.getOrDefault(fruits[r], 0) + 1);
                cur++;
                r++;
            }
            ans = Math.max(ans, cur);
            while (cnt.size() == 2) {
                cnt.put(fruits[l], cnt.get(fruits[l]) - 1);
                if (cnt.get(fruits[l]) == 0) {
                    cnt.remove(fruits[l]);
                }
                l++;
                cur--;
            }
        }
        return ans;
    }
}

File: Hash Table/0904.Fruit Into Baskets/Solution.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        l, r, n = 0, 0, len(fruits)
        ans = cur = 0
        mp = dict()
        while r < n:
            while r < n and (fruits[r] in mp or len(mp) < 2):
                mp[fruits[r]] = mp.get(fruits[r], 0) + 1
                cur += 1
                r += 1
            ans = max(ans, cur)
            while len(mp) == 2:
                mp[fruits[l]] -= 1
                if mp[fruits[l]] == 0:
                    mp.pop(fruits[l])
                cur -= 1
                l += 1
        return ans
